prov/verbs/src/fi_verbs.h

```diff
@@ -114,6 +114,7 @@
 
 #define VERBS_NO_COMP_FLAG	((uint64_t)-1)
 
+#define VRB_CM_REQ_HDR_SIZE (36)
```

Comment:
I think there's confusion as to what's happening here.

When the RDMA CM protocol is used (as determined by the service ID range), then the IB CM REQ packet must carry the RDMA CM header.  The kernel behavior here differs based on the addressing that is used by the app.  For AF_INET or AF_INET6, the kernel is responsible for formatting the RDMA CM header.  The application level private data follows that header.

For AF_IB, the change is that the kernel no longer has the information that it needs in order to format the RDMA CM header.  The only data that it fill out is the version field, which is the first byte of the RDMA CM header.  This is the 'corruption' that the commit message is referencing.  When using AF_IB, user space should fill out the RDMA CM header and pass that to the kernel.  Any application level private data should still follow that header.

The verbs provider should provide the RDMA CM header, even if AF_IB is used.  The GIDs can go in place of the IP addresses.  The application data should follow that. See this link for the RDMA CM header format:

https://github.com/linux-rdma/rdma-core/blob/master/librdmacm/cma.h#L105

struct vrb_cm_data_hdr would follow that data.

Reply:
Thanks for the detailed answer, @shefty! I learn a lot from your reviews.

I have modified the patch series to fix the issues with the RDMA CM header.
I created a new structure `vrb_rdma_cm_hdr`, which is identical to to the structure `ib_connect_hdr` in RDMA CM.
Source and destination addresses in the RDMA CM header are set based on the addresses stored in `rdma_cm_id::route.addr.{src_addr,dst_addr}`. Regarding `vrb_rdma_cm_hdr::ip_version` and `vrb_rdma_cm_hdr::port`, I set both values to 0. Is that correct?

Reply:
I think ip_version should be 6 (a GID is an IPv6 address, except that it's not), and the port should match the lower 16-bits of the SID, depending on whether you're referring to either the source or destination.